### PR TITLE
Add summation kernels

### PR DESCRIPTION
**Pull request by junjihashimoto**
This PR implements reduction-kernels.
It works correctly, but it's only about twice as fast as CPU's one.

```
$ ./build/reduce
Initializing 67108864 values
[info] Requesting adapter
[info] Requesting device
[info] Waiting for device request to end
[info] Device request ended
Start testing sum(x) on 67108864 values
Duration(CPU): 4493.9 microseconds
sum_cpu: -3248.239746
Start testing sum(x) on 67108864 values
Duration(GPU): 2175.1 microseconds
sum_gpu: -3248.113770
Success: diff = 0.125977
Computed 67108864 values of kSum(x)
[info] Context destroyed
```

### CODE REVIEW

File: experimental/kernels/reduce.cpp [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+#include <cstdio>
+#include <future>
+#include <random>
+#include <algorithm>
+#include "utils/array_utils.hpp"    // show, isclose, randn, randint
+#include "kernels.h"
+
+using namespace gpu;
+
+#define LIMITS { \
```
Comment (austinvhuang) [inline experimental/kernels/reduce.cpp:12]: this is fine as is for the experimental implementation, i wonder if there's a way we can populate this by polling system properties?

**Comment by austinvhuang** [comment]
LGTM, 2x improvement doesn't seem to bad for a laptop. Any reason not to merge yet?

**Comment by junjihashimoto** [comment]
@austinvhuang 
I set this a draft because the 2D reduction didn't work properly. 
gpt2-backward uses 2D reduction.

It works now, but it's much slower than the CPU version.
Might be ok to merge it for now.

File: experimental/kernels/reduce.cpp (hunk 1 above)
Comment (junjihashimoto) [reply experimental/kernels/reduce.cpp:12]: https://github.com/AnswerDotAI/gpu.cpp/issues/72

**Comment by junjihashimoto** [comment]
@austinvhuang Thank you!

**Comment by austinvhuang** [comment]
Thanks! Merged - will take a look at the 2D reduction.

**Pull request merged**